### PR TITLE
Users REST API

### PR DESCRIPTION
**Pull request by bbengfort**
### Scope of changes

Implements the Users management REST API.

### Type of change

- [ ] bug fix
- [x] new feature
- [ ] documentation
- [ ] other (describe)

### Acceptance criteria

Does this meet the standards and conform to the spec of our API implementation?

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ] I have updated the dependencies list
- [x] I have added new test fixtures as needed to support added tests
- [ ] I have added or updated the documentation
- [ ] Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] To the best of my ability, I believe that this PR represents a good solution to the specified problem and that it should be merged into the main code base.

### CODE REVIEW

File: pkg/store/mock/store.go [hunk 1]
```diff
@@ -165,6 +165,10 @@ func (s *Store) DeleteUser(ctx context.Context, userID ulid.ULID) error {
 	return nil
 }
 
+func (s *Store) LookupRole(ctx context.Context, role string) (*models.Role, error) {
+	return nil, nil
```
Comment (bbengfort) [inline pkg/store/mock/store.go:169]: Used to validate the role the user submits to the server.

File: pkg/store/sqlite/auth.go [hunk 2]
```diff
@@ -223,6 +224,31 @@ func (s *Store) DeleteUser(ctx context.Context, userID ulid.ULID) (err error) {
 	return tx.Commit()
 }
 
+const lookupRoleSQL = "SELECT * FROM roles WHERE title like :role LIMIT 1"
```
Comment (bbengfort) [inline pkg/store/sqlite/auth.go:227]: Note that `like` in sqlite3 is case insensitive, for postgres we would have to use `ilike`

File: pkg/store/store.go [hunk 3]
```diff
@@ -131,7 +132,9 @@ type UserStore interface {
 	RetrieveUser(ctx context.Context, emailOrUserID any) (*models.User, error)
 	UpdateUser(context.Context, *models.User) error
 	SetUserPassword(ctx context.Context, userID ulid.ULID, password string) error
+	SetUserLastLogin(ctx context.Context, userID ulid.ULID, lastLogin time.Time) error
```
Comment (bbengfort) [inline pkg/store/store.go:135]: Added this so the user's last login timestamp wouldn't get nulled out on update.

File: pkg/web/users.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+			c.JSON(http.StatusBadRequest, api.Error(api.ValidationError(nil, api.IncorrectField("role", "unknown role - specify one of admin, compliance, or observer"))))
+			return
+		}
+
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error("could not process create user request"))
+		return
+	}
+
+	// Convert the API serializer into a dtabase model
```
Comment (daniellemaxwell) [inline pkg/web/users.go:98]:
```suggestion
	// Convert the API serializer into a database model
```

**Review by daniellemaxwell — APPROVED** [review]
Nice work on making these changes. I'm sure users will greatly benefit from having this feature included in the beta release.

**Comment by ping-localhost** [comment]
Will you create an MR to update the OpenAPI specification as well? 

---

At this stage, I wonder if using https://github.com/swaggo/swag would be better, since it uses comments to automatically generate the OpenAPI definition.

**Comment by bbengfort** [comment]
> Will you create an MR to update the OpenAPI specification as well? 😅
> 
> I wonder if using https://github.com/swaggo/swag would be better, since it uses comments to automatically generate the OpenAPI definition. As in, rewriting it now is easier than later.

We are planning on updating the OpenAPI specification as well - right now we're trying to finalize the beta release for July 10th -- once the beta release is done, we can update the API docs with confidence that nothing will change. 

We have used swaggo/swag in the past, but found it didn't meet all of our requirements; when we create the API spec we also use it to generate automated tests. Normally we would included documentation and updating the OpenAPI spec (and tests) as part of our definition of done -- our development process is just a bit abnormal as we are rushing the Envoy development. Thank you so much for your patience!

**Pull request merged**